File: client/src/types/quiz.ts
import { QuizStatus, QuizMode, ExamStartMode } from './enums';

interface ExamFields {
  mode: QuizMode;
  isValidated: boolean;
  maxAttempts: number;
  startMode?: ExamStartMode | null;
  scheduledStartAt?: string | null;
  scheduledEndAt?: string | null;
  manualStartedAt?: string | null;
  joinWindowMinutes: number;
  allowFeedback: boolean;
  allowReevaluation: boolean;
  autoReevaluationQuota: boolean;
  maxReevaluationsPerStudent: number;
}

export interface QuizListDto extends ExamFields {
  id: string;
  title: string;
  description?: string;
  timeLimitMinutes: number;
  status: QuizStatus;
  passingScorePercentage: number;
  totalQuestions: number;
  createdAt: string;
  attemptsUsed: number;
}

export interface QuizDetailDto extends ExamFields {
  id: string;
  title: string;
  description?: string;
  timeLimitMinutes: number;
  status: QuizStatus;
  passingScorePercentage: number;
  categories: QuizCategoryDetailDto[];
}

export interface QuizCategoryDetailDto {
  categoryId: string;
  categoryName: string;
  questionCount: number;
  easyPercentage: number;
  mediumPercentage: number;
  hardPercentage: number;
  multipleChoicePercentage: number;
  singleChoicePercentage: number;
  trueFalsePercentage: number;
  inputPercentage: number;
}

export interface CreateQuizRequest {
  title: string;
  description?: string;
  timeLimitMinutes: number;
  passingScorePercentage: number;
  categories: {
    categoryId: string;
    questionCount: number;
    easyPercentage: number;
    mediumPercentage: number;
    hardPercentage: number;
    multipleChoicePercentage: number;
    singleChoicePercentage: number;
    trueFalsePercentage: number;
    inputPercentage: number;
  }[];
  // Exam fields
  mode: QuizMode;
  maxAttempts: number;
  startMode?: ExamStartMode | null;
  scheduledStartAt?: string | null;
  scheduledEndAt?: string | null;
  joinWindowMinutes: number;
  // Feedback / Re-evaluation
  allowFeedback: boolean;
  allowReevaluation: boolean;
  autoReevaluationQuota: boolean;
  maxReevaluationsPerStudent: number;
}

export interface UpdateQuizRequest extends CreateQuizRequest {
  id: string;
}

// Preview DTOs (teacher view with correct answers)
export interface QuizPreviewDto {
  id: string;
  title: string;
  description?: string;
  mode: string;
  isValidated: boolean;
  questions: PreviewQuestionDto[];
}

export interface PreviewQuestionDto {
  id: string;
  text?: string;
  questionType: string;
  contentType: string;
  difficultyLevel: string;
  categoryName: string;
  answers: PreviewAnswerDto[];
  media: PreviewMediaDto[];
}

export interface PreviewAnswerDto {
  id: string;
  text: string;
  isCorrect: boolean;
  orderIndex: number;
}

export interface PreviewMediaDto {
  url: string;
  fileName: string;
  mediaType: string;
}

File: src/QuizApp.Domain/Entities/Quiz.cs
using QuizApp.Domain.Common;
using QuizApp.Domain.Enums;

namespace QuizApp.Domain.Entities;

public class Quiz : BaseEntity
{
    public string Title { get; set; } = string.Empty;
    public string? Description { get; set; }
    public int TimeLimitMinutes { get; set; }
    public QuizStatus Status { get; set; } = QuizStatus.Draft;
    public int PassingScorePercentage { get; set; } = 50;
    public string CreatedById { get; set; } = string.Empty;

    // Exam mode fields
    public QuizMode Mode { get; set; } = QuizMode.Learning;
    public bool IsValidated { get; set; }
    public int MaxAttempts { get; set; } // 0 = unlimited
    public ExamStartMode? StartMode { get; set; }
    public DateTime? ScheduledStartAt { get; set; }
    public DateTime? ScheduledEndAt { get; set; }
    public DateTime? ManualStartedAt { get; set; }
    public int JoinWindowMinutes { get; set; } = 5;

    // Feedback / Re-evaluation
    public bool AllowFeedback { get; set; } = true;
    public bool AllowReevaluation { get; set; }
    // If true (default), per-student max = number of wrong answers in the attempt
    // (capped between 0 and TotalQuestions). If false, MaxReevaluationsPerStudent is used.
    public bool AutoReevaluationQuota { get; set; } = true;
    public int MaxReevaluationsPerStudent { get; set; } = 1;

    public ApplicationUser CreatedBy { get; set; } = null!;
    public ICollection<QuizCategory> QuizCategories { get; set; } = [];
    public ICollection<QuizAttempt> Attempts { get; set; } = [];
}
